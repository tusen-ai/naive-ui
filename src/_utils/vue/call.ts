type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
