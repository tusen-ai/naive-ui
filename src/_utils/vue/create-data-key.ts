export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
