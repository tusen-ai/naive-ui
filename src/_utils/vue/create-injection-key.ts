import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
