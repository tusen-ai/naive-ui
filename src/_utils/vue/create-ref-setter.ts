import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
