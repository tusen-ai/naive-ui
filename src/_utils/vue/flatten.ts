import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
