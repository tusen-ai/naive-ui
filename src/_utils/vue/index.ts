export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
