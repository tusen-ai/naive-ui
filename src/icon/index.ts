export { iconProps, NIcon } from './src/Icon'
export type { IconProps } from './src/Icon'
