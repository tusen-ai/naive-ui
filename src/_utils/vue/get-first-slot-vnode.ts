import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
