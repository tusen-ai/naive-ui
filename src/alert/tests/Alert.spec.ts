import { h } from 'vue'
import { NAlert } from '../index'
import { mount } from '@vue/test-utils'
import { NIcon } from '../../icon'
import { IosAirplane } from '@vicons/ionicons4'

describe('n-alert', () => {
  it('should work with import on demand', () => {
    mount(NAlert)
  })

  it('should have a role of "alert"', () => {
    const wrapper = mount(NAlert)
    expect(wrapper.find('.n-alert').attributes('role')).toBe('alert')
  })

  it('should add the right aria', () => {
    const wrapper = mount(NAlert)
    expect(wrapper.find('.n-alert__icon').attributes('aria-hidden')).toBe(
      'true'
    )
  })

  it('shouldnt have default title', () => {
    const wrapper = mount(NAlert)
    expect(wrapper.find('.n-alert-body__title').exists()).toBe(false)
  })

  it('should have designated title', () => {
    const title = 'sometimes naÃ¯ve'
    const wrapper = mount(NAlert, {
      props: { title }
    })
    expect(wrapper.find('.n-alert-body__title').text()).toBe(title)
  })

  it('should work with type prop', async () => {
    ;(['info', 'success', 'warning', 'error'] as const).forEach((type) => {
      const wrapper = mount(NAlert, { props: { type } })
      expect(wrapper.find('.n-alert').attributes('style')).toMatchSnapshot()
    })
  })

  it('should work with `bordered` prop', async () => {
    const wrapper = mount(NAlert)
    expect(wrapper.find('.n-alert-body--bordered').exists()).toBe(true)
    await wrapper.setProps({ bordered: false })
    expect(wrapper.find('.n-alert-body--bordered').exists()).toBe(false)
  })

  it('should work with `default` slot', () => {
    const wrapper = mount(NAlert, {
      slots: {
        default: () => 'default'
      }
    })

    expect(wrapper.find('.n-alert-body__content').exists()).toBe(true)
    expect(wrapper.find('.n-alert-body__content').text()).toBe('default')
  })

  it('should work with `icon` slot', async () => {
    const wrapper = mount(NAlert, {
      slots: {
        icon: () =>
          h(NIcon, null, {
            default: () => h(IosAirplane)
          })
      }
    })

    expect(wrapper.findComponent(NIcon).exists()).toBe(true)
  })

  it('should work with `header` slot', async () => {
    const wrapper = mount(NAlert, {
      slots: {
        header: () => 'test-header'
      }
    })

    expect(wrapper.find('.n-alert-body__title').text()).toBe('test-header')
  })

  it('shouldnt be closable by default', () => {
    const wrapper = mount(NAlert)
    expect(wrapper.find('.n-base-close.n-alert__close').exists()).toBe(false)
  })

  it('should be closable when designated', () => {
    const wrapper = mount(NAlert, { props: { closable: true } })
    expect(wrapper.find('.n-base-close.n-alert__close').exists()).toBe(true)
  })

  it('should show icon by default', () => {
    const wrapper = mount(NAlert)
    expect(wrapper.find('.n-alert__icon').exists()).toBe(true)
  })

  it('should hide icon when designated', () => {
    const wrapper = mount(NAlert, { props: { showIcon: false } })
    expect(wrapper.find('.n-alert__icon').exists()).toBe(false)
  })

  it("shouldn't closed when on-close prop returns false", async () => {
    const wrapper = mount(NAlert, {
      props: { closable: true, onClose: () => false }
    })
    const closeBtn = wrapper.find('.n-base-close.n-alert__close')
    await closeBtn.trigger('click')

    expect(wrapper.find('.n-base-close.n-alert__close').exists()).toBe(true)
  })

  it('should trigger callback when closed', async () => {
    const handleCloseClick = jest.fn()
    const handleOnAfterLeave = jest.fn()
    const wrapper = mount(NAlert, {
      props: {
        closable: true,
        onClose: handleCloseClick,
        onAfterLeave: handleOnAfterLeave
      }
    })
    const closeBtn = wrapper.find('.n-base-close.n-alert__close')
    expect(closeBtn.exists()).toBe(true)

    await closeBtn.trigger('click')
    expect(wrapper.emitted()).toHaveProperty('click')

    expect(handleCloseClick).toHaveBeenCalled()

    setTimeout(() => {
      expect(handleOnAfterLeave).toHaveBeenCalled()
    }, 0)
  })
})
