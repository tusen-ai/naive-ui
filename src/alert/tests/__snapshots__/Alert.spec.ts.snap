// Vitest Snapshot v1

exports[`n-alert > should work with type prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(237, 245, 254, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #2080f0; --n-border: 1px solid rgba(199, 223, 251, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert > should work with type prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(237, 247, 242, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #18a058; --n-border: 1px solid rgba(197, 231, 213, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert > should work with type prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(254, 247, 237, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #f0a020; --n-border: 1px solid rgba(250, 224, 181, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert > should work with type prop 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(251, 238, 241, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #d03050; --n-border: 1px solid rgba(243, 203, 211, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert should work with type prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(237, 245, 254, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #2080f0; --n-border: 1px solid rgba(199, 223, 251, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert should work with type prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(237, 247, 242, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #18a058; --n-border: 1px solid rgba(197, 231, 213, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert should work with type prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(254, 247, 237, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #f0a020; --n-border: 1px solid rgba(250, 224, 181, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;

exports[`n-alert should work with type prop 4`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-color: rgba(251, 238, 241, 1); --n-close-icon-size: 16px; --n-close-border-radius: 3px; --n-close-color-hover: rgba(0, 0, 0, .09); --n-close-color-pressed: rgba(0, 0, 0, .13); --n-close-icon-color: rgba(102, 102, 102, 1); --n-close-icon-color-hover: rgba(102, 102, 102, 1); --n-close-icon-color-pressed: rgba(102, 102, 102, 1); --n-icon-color: #d03050; --n-border: 1px solid rgba(243, 203, 211, 1); --n-title-text-color: rgb(31, 34, 37); --n-content-text-color: rgb(51, 54, 57); --n-line-height: 1.6; --n-border-radius: 3px; --n-font-size: 14px; --n-title-font-weight: 500; --n-icon-size: 24px; --n-icon-margin: 11px 8px 0 12px; --n-icon-margin-rtl: 11px 12px 0 8px; --n-close-size: 20px; --n-close-margin: 13px 14px 0 0; --n-close-margin-rtl: 13px 0 0 14px; --n-padding: 13px; --n-icon-margin-left: 12px; --n-icon-margin-right: 8px;"`;
